src/btree/bt_split.c

```diff
@@ -1572,8 +1572,8 @@ __split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *mul
             supd->ins->upd = NULL;
 
         /* Free the updates written to the data store and the history store. */
-        tmp = supd->onpage_tombstone != NULL ? &supd->onpage_tombstone : &supd->onpage_upd;
-        if (tmp != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY)) {
```

Comment:
I think we need a comment explaining the new logic of checking for the onpage_update. I don't understand why freeing a globally visible tombstone would be problematic if its on page?

Reply:
The failure is due to double free of the tombstone. If there is an `onpage_upd` marked, we truncate the update list till the point of either `onpage_tombstone` or `onpage_upd`. But in the case of globally visible tombstone, it doesn't do that and later as part of update_obsolete_check it is possible to free and also it will be freed as part of split_in_mem_final function. 

The solution could be either truncate the update list even for only `onpage_tombstone` or don't free the `onpage_tombstone` as part of the final function. I went with the later approach.

Reply:
Comment looks good thanks.

Reply:
The comment is updated to explain this scenario.